### ISSUE TITLE
cannot find spatial branch in model code

### ISSUE THREAD

**Issue by inderpreet-adapdix** [fudan-zvg/SeaFormer#10]
Hi,

I was looking into model implementation and found that Fusion_block in seaformer.py file is not returning anything in its forward pass https://github.com/fudan-zvg/SeaFormer/blob/db38fe7df1adec41ebf8125efe49184ad7396097/seaformer-cls/seaformer.py#L338-L368

also there seems to be error in its implementation: 
- global_feat is referenced before assignment,
- name InjectionMultiSum is used as the name of class while initializing super class

While debugging the forward pass for seaformer-small model I could not find the use of the Fusion block and model weights for fusion block are also not present in checkpoint. It seems like model implementation only has shared stem and context branch.

Can you please help me with these issues if I m missing something?

Thanks

**Comment by wwqq** [comment]
Hi, @inderpreet-adapdix Fusion_block (Spatial branch) is not used in the classification task. We've deleted this class. It is referenced in the segmentation code. https://github.com/fudan-zvg/SeaFormer/blob/f64a6b5a5bfc72d91fbeae9ac4f54cccd72e5fe4/seaformer-seg/mmseg/models/decode_heads/light_head.py#L69

**Comment by inderpreetsingh01** [comment]
Thanks for the reply @wwqq, I am actually using CPU and was able to get inference done on CPU for a classification model, I am not sure how to do it for the segmentation model, can you help me with it I want to evaluate seaformer-small for segmentation on CPU.

**Comment by wwqq** [comment]
The convert file is uploaded. https://github.com/fudan-zvg/SeaFormer/blob/main/convert2onnx.py
run: ```python3 convert2onnx.py <config-file> --input-img <img-dir> --shape 512 512 --checkpoint <model-ckpt>``` to convert model to ONNX.
test inference speed  in mobile devices, please refer to [TopFormer](https://github.com/hustvl/TopFormer/blob/main/tnn_runtime.md)

**Comment by inderpreetsingh01** [comment]
Hi @wwqq I followed the link you mentioned, I am doing inference for seaformer small model in jupyter notebook, I am using model SeaFormer-S_512x512_4x8_160k, I have initialized the model using config present in seaformer_small.py and loaded the weights. I am passing it an image from ADE20K dataset after resizing it to (512, 512) and result I am getting is of (64, 64). I have done bilinear interpolation to convert into size (512, 512) and mapped each class to a RGB color, the output image I am getting doesn't seem to be good and it seems image is not getting segmented. I have updated Conv2d_BN to use nn.BatchNorm2d instead of build_layer_norm.
Here is the link to notebook https://github.com/inderpreetsingh01/models/blob/master/Seaformer_segmentation_inference.ipynb
It would be helpful if you can help me with this, if I am doing something wrong here?

Thanks

**Comment by wwqq** [comment]
Hi @inderpreetsingh01 
**1**.Logits need to be upsampled first and then argmax.
<img width="762" alt="1" src="https://user-images.githubusercontent.com/59908581/223659258-bac94892-99eb-4367-83c2-24f5b2f4e88a.png">

**2**.Label_colors need to generate 150 classes not 125. You can import it from the checkpoint directly.
<img width="462" height="262" alt="1" src="https://user-images.githubusercontent.com/59908581/223661254-aefea5a8-ef71-4c45-bd4c-c9586b94da5d.jpg">

**3**.Same here, nc=150. And rgb_image should transpose not reshape.
<img width="562" height="262" alt="1" src="https://user-images.githubusercontent.com/59908581/223661151-b361deca-857b-46a2-8363-43671bdfde2b.png">

Final vis result:
<img width="262" height="262" alt="1" src="https://user-images.githubusercontent.com/59908581/223662030-96d771b9-9e14-44ff-8b6a-7fc5c3e958a3.png">

**Comment by inderpreetsingh01** [comment]
Thanks a lot, @wwqq for the clear explanation. I have one general doubt why the output seems a bit noisy, is there any way to improve it for the same checkpoint?

**Comment by speedinghzl** [comment]
Hi @inderpreetsingh01, thanks for your interest in our work.
I saw you use the simple transform as the image preprocess, it is mismatched with the preprocess used in the Seaformer training step. The segmentation result could be better if you align the preprocess, including normalization, rgb order(not sure).

<img width="966" alt="image" src="https://user-images.githubusercontent.com/4509744/223693860-e85d8fc3-1753-4af3-99f3-f9040db4d118.png">

<img width="475" alt="image" src="https://user-images.githubusercontent.com/4509744/223693785-f4c413f6-2afb-4665-bf7a-62b674a33cde.png">

**Comment by inderpreetsingh01** [comment]
Hi @speedinghzl thanks for reply, I have normalized the image now using mean=[123.675, 116.28, 103.53] and std=[58.395, 57.12, 57.375], also I was using model in training mode I converted it to eval mode and below are the output I got.

ADE_val_00001001.jpg
![image](https://user-images.githubusercontent.com/54892545/224472265-e5ec5e5d-cccb-4c15-99a9-62de0f9feffb.png)

ADE_val_00001005.jpg
![image](https://user-images.githubusercontent.com/54892545/224472331-d96211f2-4a04-488d-9736-d9fe77b2b72e.png)


It seems better than the previous one and less noisy.
Can you please confirm if is this the expected output of the seaformer small model?

**Comment by wwqq** [comment]
Hi, @inderpreetsingh01 Yes, they are the expected outputs.
